File: typescript/src/fin_data.ts
import Decimal from 'decimal.js';
import { Candle, CandleNumber, parseCandlesFromJson } from './candle';

/**
 * Error codes for FunctionArgs errors
 */
export enum ErrorCode {
  SymbolNotFound = 1,
  SourceNotFound = 2,
  TickerNotFound = 3,
  CallArgumentNotFound = 4,
  CallArgumentParseError = 5,
}

/**
 * Custom error with return code
 */
export class FunctionError extends Error {
  code: ErrorCode;

  constructor(message: string, code: ErrorCode) {
    super(message);
    this.name = 'FunctionError';
    this.code = code;
  }
}

/**
 * Represents ticker data with candles and precision
 */
export class TickersData {
  symbol: string;
  exchange: string;
  candles: CandleNumber[];
  precision: number;

  constructor(symbol: string, exchange: string, candles: CandleNumber[], precision: number) {
    this.symbol = symbol;
    this.exchange = exchange;
    this.candles = candles;
    this.precision = precision;
  }

  /**
   * Get an iterator over the candles
   */
  *getCandlesIter(): IterableIterator<CandleNumber> {
    for (const candle of this.candles) {
      yield candle;
    }
  }

  /**
   * Get the candles array
   */
  getCandles(): CandleNumber[] {
    return this.candles;
  }

  /**
   * Get an iterator over candles converted to Decimal
   */
  *getCandlesDecimalIter(): IterableIterator<Candle<Decimal>> {
    for (const candle of this.candles) {
      yield candle.toDecimal(this.precision);
    }
  }

  /**
   * Get candles as Decimal array
   */
  getCandlesDecimal(): Candle<Decimal>[] {
    return Array.from(this.getCandlesDecimalIter());
  }

  /**
   * Parse TickersData from JSON object
   */
  static fromJson(data: any): TickersData {
    return new TickersData(
      data.symbol,
      data.exchange,
      parseCandlesFromJson(data.candles),
      data.precision
    );
  }
}

/**
 * Main function arguments structure containing all input data
 */
export class FunctionArgs {
  private tickersData: Map<string, TickersData>;
  private pipedData: Map<string, string>;
  private callArguments: Map<string, string>;

  constructor(
    tickersData: Map<string, TickersData>,
    pipedData: Map<string, string>,
    callArguments: Map<string, string>
  ) {
    this.tickersData = tickersData;
    this.pipedData = pipedData;
    this.callArguments = callArguments;
  }

  

  /**
   * Get all ticker labels (keys)
   */
  getLabels(): string[] {
    return Array.from(this.tickersData.keys());
  }

  /**
   * Get candles for a specific label
   */
  getCandles(label: string): CandleNumber[] {
    const ticker = this.tickersData.get(label);
    if (!ticker) {
      throw new FunctionError(`Symbol ${label} not found`, ErrorCode.SymbolNotFound);
    }
    return ticker.candles;
  }

  /**
   * Get candles iterator for a specific label
   */
  *getCandlesIter(label: string): IterableIterator<CandleNumber> {
    const candles = this.getCandles(label);
    for (const candle of candles) {
      yield candle;
    }
  }

  /**
   * Get all pipe source keys
   */
  getPipeSources(): string[] {
    return Array.from(this.pipedData.keys());
  }

  /**
   * Get data from a specific pipe source
   */
  getDataFromPipe(source: string): string {
    const data = this.pipedData.get(source);
    if (data === undefined) {
      throw new FunctionError(`Source ${source} not found`, ErrorCode.SourceNotFound);
    }
    return data;
  }

  /**
   * Get ticker data for a specific label
   */
  getTicker(label: string): TickersData {
    const ticker = this.tickersData.get(label);
    if (!ticker) {
      throw new FunctionError(`Ticker ${label} not found`, ErrorCode.TickerNotFound);
    }
    return ticker;
  }

  /**
   * Get candles as Decimal iterator for a specific label
   * Precision is taken from the ticker
   */
  *getCandlesDecimalIter(label: string): IterableIterator<Candle<Decimal>> {
    const ticker = this.getTicker(label);
    for (const decimalCandle of ticker.getCandlesDecimalIter()) {
      yield decimalCandle;
    }
  }

  /**
   * Get candles as Decimal array for a specific label
   * Precision is taken from the ticker
   */
  getCandlesDecimal(label: string): Candle<Decimal>[] {
    return Array.from(this.getCandlesDecimalIter(label));
  }

  /**
   * Get all call arguments as a Map
   */
  getCallArguments(): Map<string, string> {
    return this.callArguments;
  }

  /**
   * Get a specific call argument and parse it
   */
  getCallArgument<T = string>(key: string, parser?: (value: string) => T): T {
    const arg = this.callArguments.get(key);
    if (arg === undefined) {
      throw new FunctionError(
        `Call argument ${key} not found`,
        ErrorCode.CallArgumentNotFound
      );
    }

    if (!parser) {
      return arg as unknown as T;
    }

    try {
      return parser(arg);
    } catch (error) {
      throw new FunctionError(
        `Call argument ${key} could not be parsed`,
        ErrorCode.CallArgumentParseError
      );
    }
  }

  /**
   * Parse FunctionArgs from JSON object
   */
  static fromJson(data: any): FunctionArgs {
    const tickersData = new Map<string, TickersData>();
    for (const [key, value] of Object.entries(data.tickers_data || {})) {
      tickersData.set(key, TickersData.fromJson(value));
    }

    const pipedData = new Map<string, string>();
    for (const [key, value] of Object.entries(data.piped_data || {})) {
      pipedData.set(key, value as string);
    }

    const callArguments = new Map<string, string>();
    for (const [key, value] of Object.entries(data.call_arguments || {})) {
      callArguments.set(key, value as string);
    }

    return new FunctionArgs(tickersData, pipedData, callArguments);
  }

  /**
   * Parse FunctionArgs from JSON string
   */
  static fromJsonString(jsonString: string): FunctionArgs {
    try {
      const data = JSON.parse(jsonString);
      return FunctionArgs.fromJson(data);
    } catch (error) {
      throw new Error(`Failed to parse FunctionArgs JSON: ${error}`);
    }
  }

  /**
   * Get the function arguments from the host environment
   */
  static get(): FunctionArgs {
    return FunctionArgs.fromJsonString(Host.inputString());
  }
}

File: rust/src/fin_data.rs
use extism_pdk::FromBytesOwned;
use extism_pdk::*;
use rust_decimal::prelude::*;
use serde::{Deserialize, de::DeserializeOwned};
use serde_json::Value;
use std::collections::HashMap;

use crate::Candle;

#[derive(Deserialize)]
pub struct TickersData {
    pub symbol: String,
    pub exchange: String,
    pub candles: Vec<Candle<f64>>,
    pub precision: i32,
}

impl TickersData {
    pub fn get_candles_iter(&self) -> impl Iterator<Item = &Candle<f64>> {
        self.candles.iter()
    }
    pub fn get_candles(&self) -> &Vec<Candle<f64>> {
        &self.candles
    }
    pub fn get_candles_decimal_iter(&self) -> impl Iterator<Item = Candle<Decimal>> {
        let precision = self.precision;
        self.candles
            .iter()
            .map(move |candle| candle.to_decimal(precision))
    }
    pub fn get_candles_decimal(&self) -> Vec<Candle<Decimal>> {
        self.get_candles_decimal_iter().collect()
    }
}

#[derive(Deserialize)]
pub struct FunctionArgs {
    tickers_data: HashMap<String, TickersData>,
    piped_data: HashMap<String, String>,
    call_arguments: HashMap<String, Value>,
}

impl FromBytesOwned for FunctionArgs {
    fn from_bytes_owned(bytes: &[u8]) -> Result<Self, extism_pdk::Error> {
        Ok(serde_json::from_slice(bytes)?)
    }
}

impl FunctionArgs {
    pub fn get_labels(&self) -> Vec<&String> {
        self.tickers_data.keys().collect()
    }

    pub fn get_candles(&self, label: &str) -> Result<&Vec<Candle<f64>>, WithReturnCode<Error>> {
        self.tickers_data
            .get(label)
            .and_then(|v| Some(&v.candles))
            .ok_or(WithReturnCode::new(
                Error::new(std::io::Error::new(
                    std::io::ErrorKind::Other,
                    format!("Symbol {} not found", label),
                )),
                1,
            ))
    }

    pub fn get_candles_iter(
        &self,
        label: &str,
    ) -> Result<impl Iterator<Item = &Candle<f64>>, WithReturnCode<Error>> {
        let candles = self.get_candles(label)?;
        Ok(candles.iter())
    }

    pub fn get_pipe_sources(&self) -> Vec<&String> {
        self.piped_data.keys().collect()
    }

    pub fn get_data_from_pipe(&self, source: &str) -> Result<&String, WithReturnCode<Error>> {
        self.piped_data.get(source).ok_or(WithReturnCode::new(
            Error::new(std::io::Error::new(
                std::io::ErrorKind::Other,
                format!("Source {} not found", source),
            )),
            2,
        ))
    }

    pub fn get_ticker(&self, label: &str) -> Result<&TickersData, WithReturnCode<Error>> {
        self.tickers_data.get(label).ok_or(WithReturnCode::new(
            Error::new(std::io::Error::new(
                std::io::ErrorKind::Other,
                format!("Ticker {} not found", label),
            )),
            3,
        ))
    }
    /// Returns the candles as Decimal, precision is taken from the ticker
    pub fn get_candles_decimal_iter(
        &self,
        label: &str,
    ) -> Result<impl Iterator<Item = Candle<Decimal>>, WithReturnCode<Error>> {
        let ticker = self.get_ticker(label)?;
        Ok(ticker.get_candles_decimal_iter())
    }
    /// Returns the candles as Decimal, precision is taken from the ticker
    pub fn get_candles_decimal(
        &self,
        label: &str,
    ) -> Result<Vec<Candle<Decimal>>, WithReturnCode<Error>> {
        Ok(self.get_candles_decimal_iter(label)?.collect())
    }

    // Returns the call arguments as a HashMap
    pub fn get_call_arguments(&self) -> &HashMap<String, Value> {
        &self.call_arguments
    }
    pub fn get_call_argument<T: DeserializeOwned>(&self, key: &str) -> Result<T, WithReturnCode<Error>> {
        let arg = self.call_arguments.get(key).ok_or(WithReturnCode::new(
            Error::new(std::io::Error::new(
                std::io::ErrorKind::Other,
                format!("Call argument {} not found", key),
            )),
            4,
        ))?;
        let res = serde_json::from_value::<T>(arg.clone()).map_err(|e| {
            WithReturnCode::new(
                Error::new(std::io::Error::new(
                    std::io::ErrorKind::Other,
                    format!("Failed to parse call argument {}: {}", key, e),
                )),
                5,
            )
        });
        match res {
            Ok(value) => Ok(value),
            Err(e) => {
                // Try to parse as string and then convert to the desired type
                if let Some(arg_str) = arg.as_str() {
                    let res_str = serde_json::from_str::<T>(arg_str).map_err(|e| {
                        WithReturnCode::new(
                            Error::new(std::io::Error::new(
                                std::io::ErrorKind::Other,
                                format!("Failed to parse call argument {} as string: {}", key, e),
                            )),
                            6,
                        )
                    });
                    match res_str {
                        Ok(value) => Ok(value),
                        Err(_) => Err(e),
                    }
                } else {
                    Err(e)
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde::{Deserialize, Serialize};
    use serde_json::json;

    #[derive(Debug, Deserialize, Serialize, PartialEq)]
    struct CustomStruct {
        name: String,
        value: i32,
    }

    fn create_test_function_args() -> FunctionArgs {
        let mut call_arguments = HashMap::new();
        
        // String value
        call_arguments.insert("string_arg".to_string(), json!("hello world"));
        
        // Integer value
        call_arguments.insert("int_arg".to_string(), json!(42));
        
        // Float value
        call_arguments.insert("float_arg".to_string(), json!(3.14));
        
        // Boolean value
        call_arguments.insert("bool_arg".to_string(), json!(true));
        
        // Object value
        call_arguments.insert("object_arg".to_string(), json!({
            "name": "test",
            "value": 100
        }));
        
        // Array value
        call_arguments.insert("array_arg".to_string(), json!([1, 2, 3, 4, 5]));

        call_arguments.insert("num_str_arg".to_string(), json!("12345"));

        call_arguments.insert("bool_str_arg".to_string(), json!("true"));
        
        call_arguments.insert("bool_str_arg_f".to_string(), json!("false"));

        call_arguments.insert("invalid_num_str_arg".to_string(), json!("not_a_number"));

        call_arguments.insert("array_str_arg".to_string(), json!("[1, 2, 3]"));

        call_arguments.insert("non_existent_arg".to_string(), json!(null));

        call_arguments.insert ("object_str_arg".to_string(), json!(r#"{"name": "test", "value": 100}"#));


        FunctionArgs {
            tickers_data: HashMap::new(),
            piped_data: HashMap::new(),
            call_arguments,
        }
    }

    #[test]
    fn test_get_call_argument_string() {
        let args = create_test_function_args();
        let result: String = args.get_call_argument("string_arg").unwrap();
        assert_eq!(result, "hello world");
    }

    #[test]
    fn test_get_call_argument_int() {
        let args = create_test_function_args();
        let result: i32 = args.get_call_argument("int_arg").unwrap();
        assert_eq!(result, 42);
    }

    #[test]
    fn test_get_call_argument_float() {
        let args = create_test_function_args();
        let result: f64 = args.get_call_argument("float_arg").unwrap();
        assert_eq!(result, 3.14);
    }

    #[test]
    fn test_get_call_argument_bool() {
        let args = create_test_function_args();
        let result: bool = args.get_call_argument("bool_arg").unwrap();
        assert_eq!(result, true);
    }

    #[test]
    fn test_get_call_argument_object() {
        let args = create_test_function_args();
        let result: CustomStruct = args.get_call_argument("object_arg").unwrap();
        assert_eq!(result, CustomStruct {
            name: "test".to_string(),
            value: 100,
        });
    }

    #[test]
    fn test_get_call_argument_array() {
        let args = create_test_function_args();
        let result: Vec<i32> = args.get_call_argument("array_arg").unwrap();
        assert_eq!(result, vec![1, 2, 3, 4, 5]);
    }

    #[test]
    fn test_get_call_argument_num_str() {
        let args = create_test_function_args();
        let result: String = args.get_call_argument("num_str_arg").unwrap();
        assert_eq!(result, "12345");

        let result_int: i32 = args.get_call_argument("num_str_arg").unwrap();
        assert_eq!(result_int, 12345);
    }

    #[test]
    fn test_get_call_argument_bool_str() {
        let args = create_test_function_args();
        let result: String = args.get_call_argument("bool_str_arg").unwrap();
        let result_f: String = args.get_call_argument("bool_str_arg_f").unwrap();
        assert_eq!(result, "true");
        assert_eq!(result_f, "false");

        let result_bool: bool = args.get_call_argument("bool_str_arg").unwrap();
        assert_eq!(result_bool, true);
        let result_bool_f: bool = args.get_call_argument("bool_str_arg_f").unwrap();
        assert_eq!(result_bool_f, false);
    }

    #[test]
    fn test_get_call_argument_invalid_num_str() {
        let args = create_test_function_args();
        let result: Result<i32, WithReturnCode<Error>> = args.get_call_argument
("invalid_num_str_arg");
        assert!(result.is_err());
    }

    #[test]
    fn test_get_call_argument_array_str() {
        let args = create_test_function_args();
        let result: Vec<i32> = args.get_call_argument("array_str_arg").unwrap();
        assert_eq!(result, vec![1, 2, 3]);
    }


    #[test]
    fn test_get_call_argument_not_found() {
        let args = create_test_function_args();
        let result: Result<String, WithReturnCode<Error>> = args.get_call_argument("non_existent_arg");
        assert!(result.is_err());
    }

    #[test]
    fn test_get_object_str_arg() {
        let args = create_test_function_args();
        let result: CustomStruct = args.get_call_argument("object_str_arg").unwrap();
        assert_eq!(result, CustomStruct {
            name: "test".to_string(),
            value: 100,
        });
    }

    #[test]
    fn test_get_call_argument_wrong_type() {
        let args = create_test_function_args();
        let result: Result<i32, WithReturnCode<Error>> = args.get_call_argument("string_arg");
        assert!(result.is_err());   
    }

    #[test]
    fn test_non_existent_argument() {
        let args = create_test_function_args();
        let result: Result<String, WithReturnCode<Error>> = args.get_call_argument("non_existent_arg");
        assert!(result.is_err());
    }
}
